linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,322 @@
+                if (result->srw_lock == NULL)
+                {
+                    LogError("Failure srw_lock_create");
+                }
+                else
+                {
+                    result->thread_count = 0;
+                    result->task_count = 0;
+                    result->quitting = 0;
+                    (void)interlocked_exchange(&result->pool_state, POOL_STATE_UNINIT);
+
+                    result->task_list = NULL;
+                    result->task_list_last = NULL;
+                    goto all_ok;
+                }
+                free(result->thread_handle_list);
+            }
+            free(result);
+            result = NULL;
+        }
+    }
+all_ok:
+    return result;
+}
+
+void threadpool_destroy(THREADPOOL_HANDLE threadpool)
+{
+    if (threadpool == NULL)
+    {
+        // do nothing
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p", threadpool);
+    }
+    else
+    {
+        interlocked_exchange(&threadpool->quitting, 1);
+
+        for (size_t index = 0; index < threadpool->list_index; index++)
+        {
+            int dont_care;
+            ThreadAPI_Join(threadpool->thread_handle_list[index], &dont_care);
+        }
+
+        free(threadpool->thread_handle_list);
+
+        srw_lock_destroy(threadpool->srw_lock);
+        free(threadpool);
+    }
+}
+
+int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCTION work_function, void* work_function_context)
+{
+    int result;
+    if (threadpool == NULL || work_function == NULL)
+    {
+        LogError("Invalid arguments: THREADPOOL_HANDLE threadpool: %p, THREADPOOL_TASK_FUNC work_function: %p", threadpool, work_function);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        THREADPOOL_TASK* task_item = malloc(sizeof(THREADPOOL_TASK));
+        if (task_item == NULL)
+        {
+            LogError("Failure allocating THREADPOOL_TASK structure");
+            result = MU_FAILURE;
+        }
+        else
+        {
+            task_item->next = NULL;
+            task_item->task_param = work_function_context;
+            task_item->task_func = work_function;
+
+            // Need to add a lock here
+            srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+            if (threadpool->task_list == NULL)
+            {
+                threadpool->task_list = task_item;
+            }
+            else
+            {
+                threadpool->task_list_last->next = task_item;
+            }
+            threadpool->task_list_last = task_item;
+            (void)interlocked_increment(&threadpool->task_count);
+
+            srw_lock_release_exclusive(threadpool->srw_lock);
+
+            // unlock
+
+            if (interlocked_add(&threadpool->pool_state, 0) == POOL_STATE_IDLE)
+            {
+                // Idle signal that we have work
+                wake_by_address_single(&threadpool->task_count);
```

Comment:
race: threadpool can get stuck because you are checking/waking on different things.

If line 233 (thread A) executes before 67 (thread B) then the work thread that should pick up pool items may not be woken until "later".

Maybe spurious wakes aren't so bad and we should just always wake here without checking the IDLE state

Reply:
Also, along with my other comment this should be a semaphore anyway and only create another thread if there aren't available threads